lib/ansible/modules/cloud/amazon/iam_server_certificate_facts.py

```diff
@@ -35,134 +42,158 @@
-
-    Kwargs:
-        name (str): The name of the server certificate.
-
-    Basic Usage:
-        >>> import boto3
-        >>> iam = boto3.client('iam')
-        >>> name = "server-cert-name"
-        >>> results = get_server_certs(iam, name)
-        {
-            "upload_date": "2015-04-25T00:36:40+00:00",
-            "server_certificate_id": "ADWAJXWTZAXIPIMQHMJPO",
-            "certificate_body": "-----BEGIN CERTIFICATE-----\nbunch of random data\n-----END CERTIFICATE-----",
-            "server_certificate_name": "server-cert-name",
-            "expiration": "2017-06-15T12:00:00+00:00",
-            "path": "/",
-            "arn": "arn:aws:iam::911277865346:server-certificate/server-cert-name"
-        }
+    :param connection: AWS boto3 glue connection
+    :param module: Ansible module
+    :param name: The name of the server certificate
+    :param path: The path of the server certificate
+    :return: IAM server certificates dict or None if not found
     """
-    results = dict()
+
+    response = dict()
+    response['certificates'] = []
+
     try:
         if name:
-            server_certs = [iam.get_server_certificate(ServerCertificateName=name)['ServerCertificate']]
+            response['certificates'].append(connection.get_server_certificate(ServerCertificateName=name))[
+                'ServerCertificate']
+        elif path:
+            server_certs = connection.list_server_certificates(PathPrefix=path)['ServerCertificateMetadataList']
         else:
-            server_certs = iam.list_server_certificates()['ServerCertificateMetadataList']
+            server_certs = connection.list_server_certificates()['ServerCertificateMetadataList']
 
-        for server_cert in server_certs:
-            if not name:
-                server_cert = iam.get_server_certificate(ServerCertificateName=server_cert['ServerCertificateName'])['ServerCertificate']
-            cert_md = server_cert['ServerCertificateMetadata']
-            results[cert_md['ServerCertificateName']] = {
-                'certificate_body': server_cert['CertificateBody'],
-                'server_certificate_id': cert_md['ServerCertificateId'],
-                'server_certificate_name': cert_md['ServerCertificateName'],
-                'arn': cert_md['Arn'],
-                'path': cert_md['Path'],
-                'expiration': cert_md['Expiration'].isoformat(),
-                'upload_date': cert_md['UploadDate'].isoformat(),
-            }
+        if not name:
+            for server_cert in server_certs:
+                response['certificates'].append(
+                    connection.get_server_certificate(ServerCertificateName=server_cert['ServerCertificateName'])[
+                        'ServerCertificate'])
 
-    except botocore.exceptions.ClientError:
-        pass
+    except botocore.exceptions.ClientError as e:
+        if e.response['Error']['Code'] == 'NoSuchEntity':
+            return response['certificates']
```

Comment:
Rather than having this and the call above in the same try/except, I think it would make sense to have a separate one for the `if not name:` block. That way, if name is not used and one of the server certificates that is initially listed is in the process of deleting this return may be premature (but it's right for the call on [line 138](https://github.com/ansible/ansible/pull/42609/files#diff-c7c2a98372cabb503e540623363950b8R138)).

Maybe something along the lines of

```
if not name:
    for server_cert in server_certs:
        try:
            # get_server_certificate and append
        except ClientError as e:
            if e.response['Error']['Code'] == 'NoSuchEntity':
                continue  # following server_certs may still exist
```